File: app/src/main/java/br/com/jonathanzanella/myexpenses/log/LogAdapter.java
package br.com.jonathanzanella.myexpenses.log;

import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import org.apache.commons.lang3.StringUtils;
import org.joda.time.DateTime;

import java.util.ArrayList;
import java.util.List;

import br.com.jonathanzanella.myexpenses.R;
import butterknife.BindView;
import butterknife.ButterKnife;

class LogAdapter extends RecyclerView.Adapter<LogAdapter.ViewHolder> {
	private final LogRepository logRepository;
	private List<Log> filteredLogs;

	public static class ViewHolder extends RecyclerView.ViewHolder {
		@BindView(R.id.row_log_indicator)
		View indicator;
		@BindView(R.id.row_log_title)
		TextView title;
		@BindView(R.id.row_log_date)
		TextView date;
		@BindView(R.id.row_log_description)
		TextView description;

		public ViewHolder(View itemView) {
			super(itemView);

			ButterKnife.bind(this, itemView);
		}

		public void setData(Log log) {
			//noinspection deprecation
			indicator.setBackgroundColor(indicator.getContext().getResources().getColor(log.getLogLevel().getColor()));
			title.setText(log.getTitle());
			date.setText(log.simpleDateFormat.format(log.getDate().toDate()));
			description.setText(log.getDescription());
		}
	}

	LogAdapter(LogRepository logRepository) {
		this.logRepository = logRepository;
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_log, parent, false);
		return new ViewHolder(v);
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		holder.setData(filteredLogs.get(position));
	}

	@Override
	public int getItemCount() {
		return filteredLogs != null ? filteredLogs.size() : 0;
	}

	public void loadData(DateTime initDate, DateTime endDate, Log.LogLevel logLevel, String filter) {
		List<Log> logs = logRepository.filter(initDate, endDate, logLevel);

		filteredLogs = new ArrayList<>();
		if(StringUtils.isNotEmpty(filter)) {
			for (Log log : logs) {
				if(StringUtils.containsIgnoreCase(log.getTitle(), filter) ||
						StringUtils.containsIgnoreCase(log.getDescription(), filter)) {
					filteredLogs.add(log);
				}
			}
		} else {
			filteredLogs = logs;
		}
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/log/LogAdapter.kt
package br.com.jonathanzanella.myexpenses.log

import android.support.v4.content.res.ResourcesCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import br.com.jonathanzanella.myexpenses.R
import kotlinx.android.synthetic.main.row_log.view.*
import org.apache.commons.lang3.StringUtils
import org.joda.time.DateTime
import java.util.*

internal class LogAdapter(private val logRepository: LogRepository) : RecyclerView.Adapter<LogAdapter.ViewHolder>() {
    private var filteredLogs: MutableList<Log>? = null

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        fun setData(log: Log) {
            itemView.indicator.setBackgroundColor(ResourcesCompat.getColor(itemView.indicator.context.resources, log.logLevel.color, null))
            itemView.title.text = log.title
            itemView.date.text = log.simpleDateFormat.format(log.date.toDate())
            itemView.description.text = log.description
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val v = LayoutInflater.from(parent.context).inflate(R.layout.row_log, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(filteredLogs!![position])
    }

    override fun getItemCount(): Int {
        return if (filteredLogs != null) filteredLogs!!.size else 0
    }

    fun loadData(initDate: DateTime, endDate: DateTime, logLevel: Log.LogLevel, filter: String) {
        val logs = logRepository.filter(initDate, endDate, logLevel)

        filteredLogs = ArrayList<Log>()
        if (StringUtils.isNotEmpty(filter)) {
            logs
                .filter { StringUtils.containsIgnoreCase(it.title, filter) || StringUtils.containsIgnoreCase(it.description, filter) }
                .forEach { filteredLogs!!.add(it) }
        } else {
            filteredLogs = logs
        }
    }
}
